Validate date-time attributes in Zeebe REST

tmetzke opened the pull request:
## Description

* Adds the `format: date-time` configuration to date properties in the OpenAPI description. This way, date properties can be properly handled by generated clients to fulfill the server's expectations.
* Validates the date-time attributes `dueDate` and `followUpDate` in the changeset
of the user task update request in the Zeebe REST API. Those attributes have to
be parseable as a date defined by [RFC 3339, section 5.6](https://www.rfc-editor.org/rfc/rfc3339#section-5.6). Possible violations are accumulated and returned
as a joined `ProblemDetail`.
* Adds a `ControllerAdvice` to the Zeebe REST gateway to ensure that we
always return a `ProblemDetail` in case of exceptions.
* Uses the details of the provided `ProblemDetail` in the `ProblemException` to provide
more information to the caller.
* Verifies that the date-time attributes `dueDate` and `followUpDate` in the changeset
of the user task update request in the Zeebe REST API are parseable as a date defined by [RFC 3339, section 5.6](https://www.rfc-editor.org/rfc/rfc3339#section-5.6).
Ensure appropriate violations are issued if they cannot be parsed accordingly.

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #16734 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [x] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

koevskinikola commented on zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RequestMapper.java:
```diff
@@ -132,16 +135,41 @@ public static Either<ProblemDetail, BrokerUserTaskUpdateRequest> toUserTaskUpdat
-    return Optional.empty();
+    if (updateRequest != null && !isEmpty(updateRequest.getChangeset())) {
+      final var changeset = updateRequest.getChangeset();
+      if (changeset.getDueDate() != null && !changeset.getDueDate().isEmpty()) {
+        try {
+          ZonedDateTime.parse(changeset.getDueDate());
+        } catch (final DateTimeParseException ex) {
+          violations.add(
+              String.format(
+                  "The provided due date '%s' cannot be parsed as a date according to RFC 3339, section 5.6.",
+                  updateRequest.getChangeset().getDueDate()));
+        }
+      }
+      if (changeset.getFollowUpDate() != null && !changeset.getFollowUpDate().isEmpty()) {
+        try {
+          ZonedDateTime.parse(changeset.getFollowUpDate());
+        } catch (final DateTimeParseException ex) {
+          violations.add(
+              String.format(
+                  "The provided follow-up date '%s' cannot be parsed as a date according to RFC 3339, section 5.6.",
+                  updateRequest.getChangeset().getFollowUpDate()));
+        }
```
:upside_down_face: Just nitpicking here, but wouldn't it be nice to extract the validation code so it's reusable for any future DateTime properties as well?

tmetzke replied:
Good catch, yeah, definitely 👍

koevskinikola commented on zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/UserTaskControllerTest.java:
```diff
@@ -432,6 +439,98 @@ public void shouldYieldBadRequestWhenUpdateTaskWithoutActionAndChanges() {
+    // when / then
+    webClient
+        .patch()
+        .uri("api/v1/user-tasks/1")
```
:x: We updated the REST API base path, it's just `/v1` now instead of `api/v1`. 

:upside_down_face: We can also extract the `v1/user-tasks` base path for user tasks into a constant in this test so that any future changes to the path are more easily adjustable.

koevskinikola commented on zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/UserTaskControllerTest.java:
```diff
@@ -432,6 +439,98 @@ public void shouldYieldBadRequestWhenUpdateTaskWithoutActionAndChanges() {
+    // when / then
+    webClient
+        .patch()
+        .uri("api/v1/user-tasks/1")
```
:x: Same as above, we need to adjust the base path of the REST API.

koevskinikola commented on zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/UserTaskControllerTest.java:
```diff
@@ -432,6 +439,98 @@ public void shouldYieldBadRequestWhenUpdateTaskWithoutActionAndChanges() {
+    // when / then
+    webClient
+        .patch()
+        .uri("api/v1/user-tasks/1")
```
:x: Same here.

koevskinikola commented on zeebe/gateway-rest/pom.xml:
```diff
@@ -231,7 +230,7 @@
               <inputSpec>${openapi.dir}/rest-api.yaml</inputSpec>
               <modelPackage>io.camunda.zeebe.gateway.protocol.rest</modelPackage>
-              <schemaMappings>ProblemDetail=org.springframework.http.ProblemDetail</schemaMappings>
+              <typeMappings>OffsetDateTime=String</typeMappings>
```
:question: Do we need to add this property in the java client as well?

tmetzke replied:
Currently, we don't need to. I am not using the date attributes of the resulting model class for user task updates from the client directly. Instead, I am setting String properties on the map of changes for the request. If we'd used those date attributes directly, we would have had to either adjust our implementation or add something like this to the client POM as well.

tmetzke replied:
For example, I am doing this in the client
```java
request.getChangeset().put(Changeset.JSON_PROPERTY_FOLLOW_UP_DATE, "");
```
where the `request.getChangeset()` is generated to be a subclass of a `HashMap` in the client.

koevskinikola reviewed (changes requested):
Looks pretty good @tmetzke :tada: 

I only have some minor feedback related to the recent change in the REST API base path and some nitpicks. Most of them will probably be partially resolved when you rebase the RP on `main`.

tmetzke commented:
After review:
* Rebased onto `main`
* Reused the user task endpoint constant in user task tests (https://github.com/camunda/zeebe/pull/16749/commits/133240fc68df68ddad6aa04e81027623389c3329)
* Extracted date validation for REST requests (https://github.com/camunda/zeebe/pull/16749/commits/8a788e92e47745f46444f7e7e6ea1bc162363acf)

koevskinikola reviewed (approved):
LGTM :tada:
